### ISSUE TITLE
AGS4: Get rid of name variants in dialog scripts etc.

### ISSUE THREAD

**Issue by messengerbag** [adventuregamestudio/ags#1273]
**Describe the bug**
Currently, some parts of AGS use variations of the entity names assigned. Specifically, dialog scripts strip off Systems-Hungarian prefixes from character names and converts to lowercase: `cRoger` becomes `roger`. If a name does not follow this convention, e.g. `Roger`, then it remains unchanged in dialog scripts. However, if there are two characters named `cRoger` and `roger`, it is not apparently possible to refer to the second one in dialog scripts.

This has a number of undesirable effects: it creates inconsistent behavior depending on whether names follow this convention; it can cause name collisions between entities _with different names_; it means that code that refers to one character can be "highjacked" by introducing/renaming another character; in some cases (e.g. importing new entities) the editor prepends the prefix automatically, which can cause duplication; and it complicates efforts to decouple tools and logic, since the full project context is necessary in order to convert dialog scripts to AGS script (see #1296); and external/third party tools for editing dialog will have to replicate this convoluted logic.

There may also be other instances of similar logic in the editor and engine, perhaps relating to other entity types (objects, GUIs, etc.).

**AGS Version**
All AGS versions to the present.

**To Reproduce**
1. Create a game project.
2. Create a character `roger`.
3. Create a dialog with the line `roger: Hello!`: the line is spoken by `roger`
4. Create a new character and name it `cRoger`: the line is now spoken by `cRoger`

**Expected behavior**
Always use and only accept the entity name exactly as assigned in the editor. If the character name is `cRoger`, dialog scripts should expect `cRoger: Hello!`.

**Comment by ericoporto** [comment]
If ones use `cRoger` instead of `Roger` on the dialog script then it's almost as useful as using `cRoger.Say` directly. The convenience of the dialog script is that most of the dialog is closer to how one would write in a script (as in movie script, not ags).

**Comment by fernewelten** [comment]
> The convenience of the dialog script is

I can't stop myself from pointing out that “convenience of the dialog script” is a contradiction in terms. At least, that's been my experience so far. 

For instance, in my last adventure Huggles needs to talk to a “bridge troll” guard. I was in prompt No. 12 of the dialogue when it became clear that another option needed to be provided after No. 4. I couldn't tack on that option at the end because the sequence of options, as presented on screen, would have been wrong.

So I moved down all the option fields from No. 12 down to No. 5 by hand, using select-all/copy/replace each time.

Obviously, the check marks next to the options (i.e., whether they were checked or not) needed to be moved as well. This is more error-prone because it __cannot__ be done through mechanical copy/paste. So I resorted to screenshotting the options before starting the moves and displaying the screenshot on another monitor after the moves, so that I could verify one-by-one that each of the marks were still set correctly. 

That left the task of changing all the option numbers.  But that at least should be a simple matter, shouldn't it? Just do a Find-all/replace on `@12`, `@11` and so on, all the way down to `@5`, right? Wrong, because the option numbers are written __without__ the preceding `@` when within code, i.e., within the indented sections of the dialogue. And you can't find-all/replace the naked numbers because they might turn up within, e.g., room co-ordinates (`if (x == 612)` contains `12`). This meant that I had to do and check carefully each instance of each replace. Which were a __lot__.

I wasted a lot of time with just this __one__ incident, and it turned out afterwards that despite all my ardent checking I had __still__ managed to put some bugs into the dialogue because of it. 

:-(

**Comment by ericoporto** [comment]
@fernewelten your comment sounds like you wanted the choices to be ordered because they are referred by number instead of a unordered identifier like a `CHOICE_MACRO_NAME`. If you are suggesting a different dialog script format, or a different improvement it may be a separate issue.

**Comment by ivan-mogilko** [comment]
> If ones use cRoger instead of Roger on the dialog script then it's almost as useful as using cRoger.Say directly. The convenience of the dialog script is that most of the dialog is closer to how one would write in a script (as in a play, not ags).

On this point I agree with @messengerbag, if character's name is cRoger then why not use it in dialogs, and why would it be more difficult to type there? I did not use dialog script myself, but I imagine it will be easier to use actual names I assigned to characters instead of remembering to lowercase them.

**Comment by ivan-mogilko** [comment]
@messengerbag please note that there's no need to use the "Bug" template when you suggest a functionality change.

**Comment by fernewelten** [comment]
@ericoporto It was an aside, certainly; I couldn't manage to sit on my fingers reading “convenience of the dialogue system”. 

In the thread I quoted above they show a presentation from AdventureX 2018 (https://www.youtube.com/watch?v=_vRfNtvFVRo).Seemingly __their__ dialogue system lets you say things like "repeat the following dialogue section until condition" or "This section will only show when the program flow has come from the user interaction so-and-so" quite naturally. Compared to what __we__ do, our `option-on` style commands seem like machine language. 

However. Our dialogue system has been around for at least a decade so I'm sure that there must already be threads suggesting fundamental impovements. So I'd rather not open another issue for fundamental improvements at this point in time; I feel that this would only clutter the bug tracker.

**Comment by messengerbag** [comment]
@ericoporto, I don't find that very persuasive, as we're only talking about a single letter, and the lack of capitalization if anything makes it less like a screenplay.

@fernewelten, some of those problems are perhaps more to do with the editor than the dialog system itself. I've never actually done extensive work in dialog script myself, but I believe that the [Dialog Designer](https://www.adventuregamestudio.co.uk/forums/index.php?topic=39568.0) tool by AJA offers a significantly better user experience (assuming it's still compatible).

It does point to another reason to remove the inconsistency: currently, having a bunch of dialog scripts means that you effectively cannot change the script-name of a character, since it would be almost impossible to update throughout. If you kept it consistent in all scripts, and if you still used a Hungarian-notation prefix, then you could fairly safely search-replace on `cRoger:`

> One way to avoid the logic would be allowing a Dialog Name property that would override the name used in the script if such property exists, then it could be used to "rename" the character script names in the dialog script name, plus this would also allow characters with spaces in their names (e.g. Roger Waters) to be used. Such property would be used only at the dialog script -> aga script conversion time and would not require any changes in the engine.

Hmm, yes, this could also be a solution. Of course, you would still need to ensure names are unique (which the current approach doesn't quite achieve).

> This happened more than once lately, perhaps the "normal/other ticket" option is not visible enough? Then we should add more formal templates.

Yes, I could see no other template, except a "blank" option.

**Comment by ericoporto** [comment]
@fernewelten I didn't mean to ask you to open an issue as a way to discourage you. @messengerbag issue is actionable with a small change but the scope of your change is a bit unclear and bigger (maybe better to discuss on the forums?). So I imagine it first requires defining what to change. It's an interesting topic and you are well equipped to actually implement a change in code. Unfortunately GitHub doesn't have functionality to move comments from one issue thread to other, so conversation in this matter may be lost.

@messengerbag

> I believe that the Dialog Designer tool by AJA offers a significantly better user experience

It does. That combined with Speech Center makes things work really well. Speech Center is a default install in all my ags installs.

> It does point to another reason to remove the inconsistency: currently, having a bunch of dialog scripts means that you effectively cannot change the script-name of a character, since it would be almost impossible to update throughout.

Overall the replace/refactor capabilities in ags are poor so I agree with this point. If it was better, the proposed change would be less of a problem.

> Hmm, yes, this could also be a solution. Of course, you would still need to ensure names are unique (which the current approach doesn't quite achieve).

To explain a bit more, what I proposed is to do as you proposed but allow overriding the character script name with a different one for the dialog scripts. Alternatively if not per character, in a table on the dialog node. 

@messengerbag @ivan-mogilko I can PR an extra feature template, it can even be pre-configured to apply a label if desired.

**Comment by ivan-mogilko** [comment]
I think we also need just "blank" template, because "blank" option is less visible with template around.

**Comment by messengerbag** [comment]
> issue is actionable with a small change but the scope of your change is a bit unclear and bigger (maybe better to discuss on the forums?). So I imagine it first requires defining what to change.

The proposed actionable change for this ticket is this: eliminate any attempt by AGS to treat entity names in any special way depending on inferred Hungarian-notation prefixes, in particular using/demanding alternative versions of character names in dialog scripts. (There may be other places where AGS tries to be clever about these prefixes; these should also be removed when discovered, but let's start here.)

If some other way of getting "friendlier" names in dialog scripts is required, that seems like it should be a separate ticket.

> Overall the replace/refactor capabilities in ags are poor so I agree with this point. If it was better, the proposed change would be less of a problem.

This doesn't follow, in my view. The replace/refactor capabilities are poor, but combined with the inconsistency it becomes even worse.

**Comment by ivan-mogilko** [comment]
> eliminate any attempt by AGS to treat entity names in any special way depending on inferred Hungarian-notation prefixes,

BTW, to illustrate the ugliness of current system:

1. Create 2 characters named "char1" and "har1", in that order.
2. Create dialog and put this:
> har1: say something!
3. Run Dialog.Start in script.

Notice that char1 is speaking.

**Comment by ericoporto** [comment]
Question on project upgrade (from 3.5.0 or 3.5.1), should it replace all dialogs names or do nothing?

**Comment by ivan-mogilko** [comment]
> Question on project upgrade (from 3.5.0 or 3.5.1), should it replace all character names in dialogs or do nothing?

Could you elaborate the question? No big changes were planned for 3.5.1.

**Comment by messengerbag** [comment]
Exactly, Ivan. (I was thinking of two characters called Charley and Harley in a game by someone not using Hungarian notation, so with scriptnames `charley` and `harley`.)

> Question on project upgrade (from 3.5.0 or 3.5.1), should it replace all character names in dialogs or do nothing?

Well, AGS4 is intended as the AGS version that drops backwards compatibility, right? So doing nothing is at least an option, since old projects are not expected to work without modification. (Of course, if someone wanted to make a conversion for it, they could, and if it used the rules as they currently are it wouldn't break anything that isn't already broken.)

**Comment by ivan-mogilko** [comment]
Sorry, looks like I misread thise question about upgrade for the first time.

We are already doing character name conversion from "dialog name" to script name using table of characters for the reference. I do not see why same function/tool can be used to upgrade dialog scripts for the new game project.
